modules/widgets/instagram/instagram.php

```diff
@@ -0,0 +1,439 @@
+			//if ( get_current_user_id() == $token->meta['user_id'] ) {
+			//	$old_token = $instance['token_id'];
+				$instance['token_id'] = (int) $_GET['token_id'];
+
+				$this->update_widget_token_id( $instance['token_id'] );
+
+			//	if ( $old_token ) {
+					//error_log( "Now delete the token: $old_token" );
+			//		$token_store->delete( array( 'type' => 'access', 'id' => $old_token ) );
+			//	}
+			//}
+		}
+		// If removing the widget's stored token ID
+		elseif ( $instance['token_id'] && isset( $_GET['instagram_widget_id'] ) && $_GET['instagram_widget_id'] == $this->number && ! empty( $_GET['instagram_widget'] ) && 'remove_token' == $_GET['instagram_widget'] ) {
+			if ( empty( $_GET['nonce'] ) || ! wp_verify_nonce( $_GET['nonce'], 'instagram-widget-remove-token-' . $this->number . '-' . $instance['token_id'] ) )
+				wp_die( __( 'Missing or invalid security nonce.', 'wpcom-instagram-widget' ) );
+
+			Keyring::init()->get_token_store()->delete( array( 'type' => 'access', 'id' => $instance['token_id'] ) );
+			$instance['token_id'] = $this->defaults['token_id'];
+
+			$this->update_widget_token_id( $instance['token_id'] );
+		}
+		// If a token ID is stored, make sure it's still valid
+		elseif ( $instance['token_id'] ) {
+			//$token = Keyring::init()->get_token_store()->get_token( array( 'type' => 'access', 'id' => $instance['token_id'] ) );
+
+			//if ( ! $token ) {
+			//	$instance['token_id'] = $this->defaults['token_id'];
+
+				$this->update_widget_token_id( $instance['token_id'] );
+			//}
+		}
+
+		// No connection, or a legacy API token? Display a connection link.
+		if ( ! $instance['token_id'] ) {
+			//$connect_url = add_query_arg( array( 'instagram_widget_id' => $this->number, 'is_customizer' => is_customize_preview() ), wpcom_keyring_get_connect_url( 'instagram_basic_display', 'instagram-widget' ) );
+
+			//echo '<p>' . __( '<strong>Important: You must first click Publish to activate this widget <em>before</em> connecting your account.</strong> After saving the widget, click the button below to authorize your Instagram account.', 'wpcom-instagram-widget' ) . '</p>';
+			//echo '<p style="text-align:center"><a class="button-primary" target="_top" href="' . esc_url( $connect_url ) . '">' . __( 'Authorize Instagram Access', 'wpcom-instagram-widget' ) . '</a></p>';
+			//echo '<p><small>' . sprintf( __( 'Having trouble? Try <a href="%s" target="_blank">logging into the correct account</a> on Instagram.com first.', 'wpcom-instagram-widget' ), 'https://instagram.com/accounts/login/' ) . '</small></p>';
+
+
+			$jetpack_blog_id = Jetpack::get_option( 'id' );
+			$response = Client::wpcom_json_api_request_as_user(
+				sprintf( '/sites/%d/external-services', $jetpack_blog_id )
+			);
+			$body = json_decode( $response['body'] );
+			$connect_URL = $body->services->instagram->connect_URL;
+			$url = add_query_arg(
+				array(
+					'siteurl' => site_url() . '/wp-admin/widgets.php',
```

Comment:
this should be modified to account for redirect from either widgets page or the customizer